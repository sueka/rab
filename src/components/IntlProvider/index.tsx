import * as React from 'react'
import { IntlProvider as OriginalIntlProvider } from 'react-intl'

export type StateProps = Pick<OriginalIntlProvider.Props, 'locale' | 'formats' | 'messages'>

type Props =
  & StateProps

// TODO: intl context でない要素を re-render しないようにする。 https://github.com/formatjs/react-intl/issues/234#issuecomment-163366518 によると現時点では難しいらしい。
// cf. https://github.com/formatjs/react-intl/issues/371#issuecomment-275703796
const IntlProvider: React.FunctionComponent<Props> = (props) => (
  <OriginalIntlProvider key={ props.locale } { ...props } />
)

export default IntlProvider
