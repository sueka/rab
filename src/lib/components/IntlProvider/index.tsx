import React from 'react'
import { IntlConfig, IntlProvider as OriginalIntlProvider } from 'react-intl'

import IntlProviderContext from '~/lib/contexts/IntlProviderContext'
import { Tag } from '~/lib/languageNameSolver'

type DefaultIntlConfig = Pick<IntlConfig, 'formats' | 'messages' | 'timeZone' | 'textComponent' | 'defaultLocale' | 'defaultFormats' | 'onError'>

type OriginalIntlProviderProps = Alt.Omit<IntlConfig, keyof DefaultIntlConfig> & Partial<DefaultIntlConfig>

export interface OwnProps extends Alt.Omit<OriginalIntlProviderProps, 'locale' | 'formats' | 'messages'> {
  availableLocales: Tag[]
}

export type StateProps = Pick<OriginalIntlProviderProps, 'locale' | 'formats' | 'messages'>

type Props =
  & OwnProps
  & StateProps

// NOTE: key が無い場合、 FormattedMessage 等は re-render されるが、 useIntl の結果は更新されない。
// TODO: intl context でない要素を re-render しないようにする。 https://github.com/formatjs/react-intl/issues/234#issuecomment-163366518 によると現時点では難しいらしい。
// cf. https://github.com/formatjs/react-intl/issues/371#issuecomment-275703796
const IntlProvider: React.FunctionComponent<Props> = ({ availableLocales, ...props }) => (
  <IntlProviderContext.Provider value={ { availableLocales } }>
    <OriginalIntlProvider key={ props.locale } textComponent={ React.Fragment } { ...props } />
  </IntlProviderContext.Provider>
)

export default IntlProvider
