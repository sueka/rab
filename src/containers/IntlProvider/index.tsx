import { connect } from 'react-redux'

import { State } from '../../redux'
import IntlProvider, { StateProps } from '../../components/IntlProvider'

const mapStateToProps = ({ localeSelector: { locale, messages } }: State): StateProps => ({
  locale,
  messages,
})

export default connect(mapStateToProps)(IntlProvider)
