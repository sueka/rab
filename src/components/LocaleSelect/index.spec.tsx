import { fireEvent, render } from '@testing-library/react'
import React from 'react'

import IntlProvider from '~/lib/components/IntlProvider'
import { LocaleSelect } from '.'

test('LocaleSelect', () => {
  const selectLocale = jest.fn()

  const { container, getByTestId } = render(
    <IntlProvider availableLocales={ ['en', 'ja'] } locale="en">
      <LocaleSelect
        locale="en"
        selectLocale={ selectLocale }
      />
    </IntlProvider>
  )

  expect(container.firstChild).toMatchSnapshot()

  fireEvent.change(getByTestId('localeSelect'), { target: { value: 'ja' }})
  expect(selectLocale).toHaveBeenCalledTimes(1)
  expect(selectLocale).toHaveBeenCalledWith('ja')
})
