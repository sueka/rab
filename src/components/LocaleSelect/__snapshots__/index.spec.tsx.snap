// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocaleSelect 1`] = `
<div
  class="MuiFormControl-root"
>
  <label
    class="MuiFormLabel-root MuiInputLabel-root InputLabel MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
    data-shrink="true"
    for="00000000-0000-4000-0000-000000000000"
  >
    <bdi>
      Languages
    </bdi>
  </label>
  <div
    class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
  >
    <div
      aria-haspopup="listbox"
      aria-labelledby="00000000-0000-4000-0000-000000000000"
      class="MuiSelect-root MuiSelect-select Select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
      id="00000000-0000-4000-0000-000000000000"
      role="button"
      tabindex="0"
    >
      English
    </div>
    <input
      aria-hidden="true"
      class="MuiSelect-nativeInput"
      data-testid="localeSelect"
      tabindex="-1"
      value="en"
    />
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSelect-icon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M7 10l5 5 5-5z"
      />
    </svg>
  </div>
</div>
`;
